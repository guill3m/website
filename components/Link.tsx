import { ReactNode } from 'react'
import NextLink from 'next/link'

type LinkProps = {
  className?: string,
  children: ReactNode,
  href: string|null,
  target?: string,
}

export default function Link ({
  children,
  href,
  target,
  ...rest
}: LinkProps) {
  if (!href) {
    return (
      <span {...rest}>{children}</span>
    )
  }

  if (typeof href === 'string' && href.startsWith('https://')) {
    return (
      <a
        href={href}
        rel='noopener noreferrer nofollow'
        target={target || '_blank'}
        {...rest}
      >
        {children}
      </a>
    )
  }

  return (
    <NextLink href={href} target={target} {...rest}>
      {children}
    </NextLink>
  )
}
