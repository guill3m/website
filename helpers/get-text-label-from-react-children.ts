import { Children, type ReactNode } from 'react'

export default function getTextLabelFromReactChildren(
  children: ReactNode | ReactNode[],
): string {
  let label = ''

  Children.map(children, (child) => {
    if (typeof child === 'string') {
      label += child
    }
  })

  return label.trim()
}
