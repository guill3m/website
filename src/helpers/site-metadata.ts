const siteMetadata = {
	author: {
		name: 'Guillem Andreu',
		social: {
			bluesky: 'guill3m.me',
			github: 'guill3m',
			lastfm: 'guill3m',
			linkedin: 'guillemandreu',
		},
	},
	host: 'www.guillemandreu.com',
}

export default siteMetadata
