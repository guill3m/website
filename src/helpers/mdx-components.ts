import Link from '../components/link.astro'

export const components = { a: Link }
