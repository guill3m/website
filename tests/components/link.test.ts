import { describe, expect, test } from 'vitest'
import { renderAstroComponent } from '../helpers'

import Link from '../../src/components/link.astro'

describe('Link component', () => {
	describe('internal link', () => {
		test('default', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: 'Internal link',
				},
				props: {
					href: '/test/',
				},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLAnchorElement)
			expect(linkElement).toHaveTextContent('Internal link')
			expect(linkElement).not.toHaveAttribute('target')
			expect(linkElement).not.toHaveAttribute('rel')
			expect(linkElement).toHaveAccessibleName('Internal link')
		})

		test('in a new tab', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: 'Internal link in new tab',
				},
				props: {
					href: '/test/',
					target: '_blank',
				},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLAnchorElement)
			expect(linkElement).toHaveTextContent('Internal link in new tab')
			expect(linkElement).toHaveAttribute('target', '_blank')
			expect(linkElement).not.toHaveAttribute('rel')
			expect(linkElement).toHaveAccessibleName(
				'Internal link in new tab (opens in a new tab)',
			)
		})

		test('with rel', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: 'Internal link',
				},
				props: {
					href: '/test/',
					rel: 'nofollow',
				},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLAnchorElement)
			expect(linkElement).toHaveTextContent('Internal link')
			expect(linkElement).not.toHaveAttribute('target')
			expect(linkElement).toHaveAttribute('rel', 'nofollow')
			expect(linkElement).toHaveAccessibleName('Internal link')
		})

		test('with additional attributes', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: 'Internal link',
				},
				props: {
					class: 'link',
					'data-testid': 'test internal',
					id: 'the-link',
					href: '/test/',
				},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLAnchorElement)
			expect(linkElement).toHaveTextContent('Internal link')
			expect(linkElement).not.toHaveAttribute('target')
			expect(linkElement).not.toHaveAttribute('rel')
			expect(linkElement).toHaveAttribute('data-testid', 'test internal')
			expect(linkElement).toHaveAttribute('id', 'the-link')
			expect(linkElement).toHaveClass('link')
			expect(linkElement).toHaveAccessibleName('Internal link')
		})

		test('with HTML content', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: '<em>Internal</em> link',
				},
				props: {
					href: '/test/',
				},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLAnchorElement)
			expect(linkElement).toHaveTextContent('Internal link')
			expect(linkElement).not.toHaveAttribute('target')
			expect(linkElement).not.toHaveAttribute('rel')
			expect(linkElement).toHaveAccessibleName('Internal link')
			expect(linkElement).toContainHTML('<em>Internal</em> link')
		})
	})

	describe('external link', () => {
		test('default', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: 'External link',
				},
				props: {
					href: 'https://www.test.com',
				},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLAnchorElement)
			expect(linkElement).toHaveTextContent('External link')
			expect(linkElement).toHaveAttribute('target', '_blank')
			expect(linkElement).toHaveAttribute('rel', 'noopener')
			expect(linkElement).toHaveAccessibleName(
				'External link (opens in a new tab)',
			)
		})

		test('with rel', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: 'External link',
				},
				props: {
					href: 'https://www.test.com',
					rel: 'nofollow',
				},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLAnchorElement)
			expect(linkElement).toHaveTextContent('External link')
			expect(linkElement).toHaveAttribute('target', '_blank')
			expect(linkElement).toHaveAttribute('rel', 'nofollow noopener')
			expect(linkElement).toHaveAccessibleName(
				'External link (opens in a new tab)',
			)
		})

		test('with additional attributes', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: 'External link',
				},
				props: {
					class: 'link',
					'data-testid': 'test external',
					id: 'the-link',
					href: 'https://www.test.com',
				},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLAnchorElement)
			expect(linkElement).toHaveTextContent('External link')
			expect(linkElement).toHaveAttribute('target', '_blank')
			expect(linkElement).toHaveAttribute('rel', 'noopener')
			expect(linkElement).toHaveAttribute('data-testid', 'test external')
			expect(linkElement).toHaveAttribute('id', 'the-link')
			expect(linkElement).toHaveClass('link')
			expect(linkElement).toHaveAccessibleName(
				'External link (opens in a new tab)',
			)
		})

		test('with overwritten aria-label', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: 'External link',
				},
				props: {
					ariaLabel: 'Link that opens in a new tab',
					href: 'https://www.test.com',
				},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLAnchorElement)
			expect(linkElement).toHaveTextContent('External link')
			expect(linkElement).toHaveAttribute('target', '_blank')
			expect(linkElement).toHaveAttribute('rel', 'noopener')
			expect(linkElement).toHaveAccessibleName('Link that opens in a new tab')
		})

		test('with HTML content', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: '<em>External</em> link',
				},
				props: {
					href: 'https://www.test.com',
				},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLAnchorElement)
			expect(linkElement).toHaveTextContent('External link')
			expect(linkElement).toHaveAttribute('target', '_blank')
			expect(linkElement).toHaveAttribute('rel', 'noopener')
			expect(linkElement).toHaveAccessibleName(
				'External link (opens in a new tab)',
			)
			expect(linkElement).toContainHTML('<em>External</em> link')
		})
	})

	describe('without href', () => {
		test('default', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: 'Link without href',
				},
				props: {},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLSpanElement)
			expect(linkElement).toHaveTextContent('Link without href')
		})

		test('with additional attributes', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: 'Link without href',
				},
				props: {
					class: 'link',
					'data-testid': 'test non-href',
					id: 'the-link',
				},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLSpanElement)
			expect(linkElement).toHaveTextContent('Link without href')
			expect(linkElement).toHaveAttribute('data-testid', 'test non-href')
			expect(linkElement).toHaveAttribute('id', 'the-link')
			expect(linkElement).toHaveClass('link')
		})

		test('with HTML content', async () => {
			const result = await renderAstroComponent(Link, {
				slots: {
					default: 'Link <em>without href</em>',
				},
				props: {},
			})
			const linkElement = result.firstChild

			expect(result).toMatchSnapshot()

			expect(linkElement).toBeInstanceOf(HTMLSpanElement)
			expect(linkElement).toHaveTextContent('Link without href')
			expect(linkElement).toContainHTML('Link <em>without href</em>')
		})
	})
})
