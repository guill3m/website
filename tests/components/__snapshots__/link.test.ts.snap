// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link component > external link > default 1`] = `
<DocumentFragment>
  <a
    aria-label="External link (opens in a new tab)"
    href="https://www.test.com"
    rel="noopener"
    target="_blank"
  >
    External link
  </a>
</DocumentFragment>
`;

exports[`Link component > external link > with HTML content 1`] = `
<DocumentFragment>
  <a
    aria-label="External link (opens in a new tab)"
    href="https://www.test.com"
    rel="noopener"
    target="_blank"
  >
    <em>
      External
    </em>
     link
  </a>
</DocumentFragment>
`;

exports[`Link component > external link > with additional attributes 1`] = `
<DocumentFragment>
  <a
    aria-label="External link (opens in a new tab)"
    class="link"
    data-testid="test external"
    href="https://www.test.com"
    id="the-link"
    rel="noopener"
    target="_blank"
  >
    External link
  </a>
</DocumentFragment>
`;

exports[`Link component > external link > with overwritten aria-label 1`] = `
<DocumentFragment>
  <a
    aria-label="Link that opens in a new tab"
    href="https://www.test.com"
    rel="noopener"
    target="_blank"
  >
    External link
  </a>
</DocumentFragment>
`;

exports[`Link component > external link > with rel 1`] = `
<DocumentFragment>
  <a
    aria-label="External link (opens in a new tab)"
    href="https://www.test.com"
    rel="nofollow noopener"
    target="_blank"
  >
    External link
  </a>
</DocumentFragment>
`;

exports[`Link component > internal link > default 1`] = `
<DocumentFragment>
  <a
    href="/test/"
  >
    Internal link
  </a>
</DocumentFragment>
`;

exports[`Link component > internal link > in a new tab 1`] = `
<DocumentFragment>
  <a
    aria-label="Internal link in new tab (opens in a new tab)"
    href="/test/"
    target="_blank"
  >
    Internal link in new tab
  </a>
</DocumentFragment>
`;

exports[`Link component > internal link > with HTML content 1`] = `
<DocumentFragment>
  <a
    href="/test/"
  >
    <em>
      Internal
    </em>
     link
  </a>
</DocumentFragment>
`;

exports[`Link component > internal link > with additional attributes 1`] = `
<DocumentFragment>
  <a
    class="link"
    data-testid="test internal"
    href="/test/"
    id="the-link"
  >
    Internal link
  </a>
</DocumentFragment>
`;

exports[`Link component > internal link > with rel 1`] = `
<DocumentFragment>
  <a
    href="/test/"
    rel="nofollow"
  >
    Internal link
  </a>
</DocumentFragment>
`;

exports[`Link component > without href > default 1`] = `
<DocumentFragment>
  <span>
    Link without href
  </span>
</DocumentFragment>
`;

exports[`Link component > without href > with HTML content 1`] = `
<DocumentFragment>
  <span>
    Link 
    <em>
      without href
    </em>
  </span>
</DocumentFragment>
`;

exports[`Link component > without href > with additional attributes 1`] = `
<DocumentFragment>
  <span
    class="link"
    data-testid="test non-href"
    id="the-link"
  >
    Link without href
  </span>
</DocumentFragment>
`;
